modules/finality-verifier/src/lib.rs

```diff
@@ -175,32 +322,207 @@ pub mod pallet {
 		FailedToWriteHeader,
 		/// There are too many requests for the current window to handle.
 		TooManyRequests,
+		/// The header being imported is on a fork which is incompatible with the current chain.
+		///
+		/// This can happen if we try and import a finalized header at a lower height than our
+		/// current `best_finalized` header.
+		ConflictingFork,
+		/// The scheduled authority set change found in the header is unsupported by the pallet.
+		///
+		/// This is the case for non-standard (e.g forced) authority set changes.
+		UnsupportedScheduledChange,
+		/// The pallet has already been initialized.
+		AlreadyInitialized,
+		/// All pallet operations are halted.
+		Halted,
+	}
+
+	/// Import the given header to the pallet's storage.
+	///
+	/// This function will also check if the header schedules and enacts authority set changes,
+	/// updating the current authority set accordingly.
+	///
+	/// Note: This function assumes that the given header has already been proven to be valid and
+	/// finalized. Using this assumption it will write them to storage with minimal checks. That
+	/// means it's of great importance that this function *not* called with any headers whose
+	/// finality has not been checked, otherwise you risk bricking your bridge.
+	pub(crate) fn import_header<T: Config>(header: BridgedHeader<T>) -> Result<(), sp_runtime::DispatchError> {
+		let best_finalized = <ImportedHeaders<T>>::get(<BestFinalized<T>>::get()).expect(
+			"In order to reach this point the bridge must have been initialized. Therefore
+			`ImportedHeaders` must contain an entry for `BestFinalized`.",
+		);
+
+		// We do a quick check here to ensure that our header chain is making progress and isn't
+		// "travelling back in time" (which would be indicative of something bad, e.g a hard-fork).
+		ensure!(best_finalized.number() < header.number(), <Error<T>>::ConflictingFork);
```

Comment:
I wonder if it makes more sense to verify these simple requirements before verifying the justification (which might be quite heavy).

Reply:
Yeah maybe. I wanted to keep as much of the "new" code in `import_header` as opposed to adding it into `submit_finality_proof` for this PR, but maybe for this small stuff it's not a big deal